### PR TITLE
RamTransformer: Hoisting RamAggregates up the loop nest

### PR DESCRIPTION
**Pull request by dcol97**
Implemented a transformer to move RamAggregates as far up the loop nest as possible.

I'm not sure how exhaustive the test cases are in testing this.

To achieve this, I had to extend RamLevelAnalysis to handle RamOperations (and missing Expressions and Conditions).

This relates to issue #992.

### CODE REVIEW

File: src/RamTransforms.cpp [hunk 1]
```diff
@@ -554,6 +554,190 @@ bool TupleIdTransformer::reorderOperations(RamProgram& program) {
...
+
+        bool added = false;
+
+        // Adding back the aggregate if one was removed before
+        std::function<std::unique_ptr<RamNode>(std::unique_ptr<RamNode>)> aggAdder =
+                [&](std::unique_ptr<RamNode> node) -> std::unique_ptr<RamNode> {
+            if (auto* search = dynamic_cast<RamSearch*>(node.get())) {
+                if (!added && search->getTupleId() == currLevel - 1) {
+                    added = true;
+                    RamNestedOperation* op = dynamic_cast<RamNestedOperation*>(search);
```
Comment (mmcgr) [inline src/RamTransforms.cpp]: Use auto rather than duplicating the type on the same line

File: src/RamTransforms.cpp [hunk 2]
```diff
@@ -554,6 +554,190 @@ bool TupleIdTransformer::reorderOperations(RamProgram& program) {
...
+                    // are aggregates, then we do not transform
+                    bool allAggragates = true;
+                    for (int i = rla->getLevel(agg) + 1; i <= agg->getTupleId(); i++) {
+                        if (!(std::find(aggIds.begin(), aggIds.end(), i) != aggIds.end())) {
+                            currLevel = i - 1;
+                            allAggragates = false;
+                            break;
+                        }
+                    }
+                    if (!allAggragates) {
```
Comment (azreika) [inline src/RamTransforms.cpp]: Aggragates -> Aggregates (and elsewhere)

Comment (mmcgr) [reply src/RamTransforms.cpp]: Oh, I didn't see this before I merged. You should have made it a review and requested changes  Spelling PR!

Comment (dcol97) [reply src/RamTransforms.cpp]: This is also my bad - Bernhard wanted the logic of the transformer to be simplified a bit so I should have prevented merging. Will cover these two things in another PR.

**Pull request merged**